src/raft.c

```diff
@@ -115,19 +115,170 @@ void entryAttachRaftReq(RedisRaftCtx *rr, raft_entry_t *entry, RaftReq *req)
 
 /* ------------------------------------ Log Execution ------------------------------------ */
 
+bool isSharding(RedisRaftCtx *rr) {
+    return rr->sharding_info->is_sharding;
+}
+
+typedef enum KeysStatus{
+    NoKeys,
+    AllExist,
+    SomeExist,
+    NoneExist,
+} KeysStatus;
+
+KeysStatus validateKeyExistence(RedisRaftCtx *rr, RaftRedisCommandArray *cmds) {
+    int total_keys = 0;
+    int found = 0;
+
+    for (int i = 0; i < cmds->len; i++) {
+        RaftRedisCommand *cmd = cmds->commands[i];
+
+        /* Iterate command keys */
+        int num_keys = 0;
+        int *keyindex = RedisModule_GetCommandKeys(rr->ctx, cmd->argv, cmd->argc, &num_keys);
+        total_keys += num_keys;
+
+        for (int j = 0; j < num_keys; j++) {
+            found += RedisModule_KeyExists(rr->ctx, cmd->argv[keyindex[j]]);
+        }
+        RedisModule_Free(keyindex);
+
+        /* shortcut as we know the result is now SomeExist */
+        if (found != 0 && found != total_keys) {
+            return SomeExist;
+        }
+    }
+
+    if (found != total_keys) {
+        return (found == 0) ? NoneExist : SomeExist;
+    }
+
+    return (total_keys == 0) ? NoKeys : AllExist;
+}
+
+/* figure out the "owner shardgroup"
+ *
+ * the owner shardgroup is either
+ *
+ * 1) a shardgroup that owns the slot as a stable slot
+ * 2) a shardgroup that owns the slot as a migrating slot.  by definition is this shardgroup is local, then it
+ *    can't also be importing (a single RedisRaft cluster cannot be both importing and migrating the same slot
+ * 3) a shardgroup marked as local (i.e. corresponding to this cluster) that owns the slot as an importing slot and
+ *    has a RaftRedisCommandArray marked as asking
+ */
+
+ShardGroup * GetCommandShardGroup(RedisRaftCtx *rr, RaftRedisCommandArray *cmds)
+{
+    ShardGroup *sg = rr->sharding_info->stable_slots_map[cmds->slot];
+    if (sg != NULL) {
+        return sg;
+    }
+
+    sg = rr->sharding_info->migrating_slots_map[cmds->slot];
+    if (sg && sg->local) {
+        return sg;
+    }
+
+    if (cmds->asking) {
+        ShardGroup *isg = rr->sharding_info->importing_slots_map[cmds->slot];
+        if (isg && isg->local) {
+            return isg;
+        }
+    }
+
+    return sg;
+}
+
+RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx, RaftRedisCommandArray *cmds) {
+    /* Make sure hash slot is mapped and handled locally. */
+    ShardGroup *sg = GetCommandShardGroup(rr, cmds);
+    if (!sg) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERDOWN Hash slot is not served");
+        }
+        return RR_ERROR;
+    }
+
+    /* this is slightly questionable.
+     * I'm keen to argue that all slots belongingto a single shardgroup have to have the same slot type
```

Comment:
hmm, I was thinking (now incorrectly) that a single redisraft cluster could be part of more than one shardgroup (i.e. can be shardgroup a and shardgroup b with node x in both cases).  Doesn't work because the shardgroup id is matched against dbid.  so need to rework this.

Reply:
modified code

```
    SlotRangeType slot_type = SLOTRANGE_TYPE_UNDEF;
    
    for (size_t i = 0; i < sg->slot_ranges_num; i++) {
        if (sg->slot_ranges[i].start_slot <= cmds->slot && cmds->slot <= sg->slot_ranges[i].end_slot) {
            slot_type = sg->slot_ranges[i].type;
            break;
        } 
    }
    
    if (slot_type == SLOTRANGE_TYPE_UNDEF) {
        if (reply_ctx) {
            RedisModule_ReplyWithError(reply_ctx, "ERR internal error, couldn't associate a shardgroup slot to this request");
        }
        return RR_ERROR;
    }
```